openpype/tools/workfiles/window.py

```diff
@@ -386,25 +384,25 @@ def _create_workfile_doc(self, filepath):
         return workfile_doc
 
     def refresh(self):
-        # Refresh asset widget
-        self.assets_widget.refresh()
+        """
+        Refresh asset widget,
+        which will refresh the task widget,
+        which will refresh the file widget
+        """
 
-        self._on_task_changed()
+        self.assets_widget.refresh()
 
     def set_context(self, context):
-        self._context_to_set = context
-        self._set_context_timer.start()
-
-    def _on_context_set_timeout(self):
-        if self._context_to_set is None:
-            self._set_context_timer.stop()
+        # type: (dict[str, Any]) -> None
+        if context is None:
             return
 
         if self.assets_widget.refreshing:
-            return
+            def _set_context():
+                self.set_context(context)
+
+            return QtCore.QTimer.singleShot(100, _set_context)
```

Comment:
These objects (function and timer) can be garbage collected before execution in older python/pyside.

Reply:
> Asset name filtering is missing and both asset and task widgets have refresh button which does not seem right.

I will resolve the ui layout so it looks and functions as before ( sloppy of me to not double check this ) but have the new components in place, just hidden.
The intention for a refresh button on tasks is to allow a localised refresh for the active asset, avoiding having to refresh the whole project which is much more expensive, especially for larger projects and remote setups. This was a request from our artists as tasks get added/updated in Jira arbitrarily.
If this doesn't make sense for some reason or limitation I haven't been able to discern from the code then that is fair enough and can be hidden permanently. 

> These objects (function and timer) can be garbage collected before execution in older python/pyside.

Ah interesting - I had wondered as to the fixed timer approach. I made this modification to test if the timer was causing the extra refreshes (it wasn't) but left it in as it was more concise. I will revert it to the previous setup to avoid the backwards compatibility issues.

Reply:
> If this doesn't make sense for some reason or limitation I haven't been able to discern from the code then that is fair enough and can be hidden permanently.

I would say it's better to have single refresh button per tool. @mkolar opinions?

Reply:
weird - something is going wrong as the asset filtering is visible for me:

![image](https://user-images.githubusercontent.com/7100563/200288805-e292a645-b7b2-4930-9a6d-36696af52751.png)

Are you able to see anything that would be hiding it for you?

Reply:
> I would say it's better to have single refresh button per tool. mkolar opinions?

Personally I agree - this double refresh button feels too cluttered and confusing. Rather have one refresh that force refreshes all.

Reply:
As mentioned, this was a request by my artists - so there is a valid use case.
I guess this is a case of form vs function?

Whilst this is on going, I am pushing updates which will include the task widget refresh button being hidden by default (for now).